sdks/java/io/kafka/src/main/java/org/apache/beam/sdk/io/kafka/KafkaUnboundedSource.java

```diff
@@ -100,8 +102,31 @@ public List<KafkaUnboundedSource<K, V>> split(int desiredNumSplits, PipelineOpti
         }
       }
     } else {
-      for (TopicPartition p : partitions) {
-        Lineage.getSources().add("kafka", ImmutableList.of(bootStrapServers, p.topic()));
+      try (Consumer<?, ?> consumer = spec.getConsumerFactoryFn().apply(spec.getConsumerConfig())) {
+        Map<String, List<Integer>> topicsAndPartitions = new HashMap<>();
+        try {
+          for (TopicPartition p : partitions) {
+            if (!topicsAndPartitions.containsKey(p.topic())) {
+              List<Integer> partitionsList = new ArrayList<>();
+              List<PartitionInfo> partitionInfoList = consumer.partitionsFor(p.topic());
+              for (PartitionInfo info : partitionInfoList) {
+                partitionsList.add(info.partition());
+              }
+              topicsAndPartitions.put(p.topic(), partitionsList);
+            }
+            checkState(
+                topicsAndPartitions.containsKey(p.topic())
+                    && topicsAndPartitions.get(p.topic()).contains(p.partition()),
+                "Partition "
+                    + p.partition()
+                    + " does not exist for topic "
+                    + p.topic()
+                    + ". Please check Kafka configuration.");
+            Lineage.getSources().add("kafka", ImmutableList.of(bootStrapServers, p.topic()));
+          }
+        } catch (KafkaException e) {
+          LOG.warn("Unable to access cluster. Skipping fail fast checks.");
+        }
```

Comment:
This can be simplified by first constructing a `Map<String, Set<Integer>>` of provided `TopicPartition`s and then iterating over `topicsAndPartitions.entrySet()`.